text/0000-route-data-loading.md

```diff
@@ -0,0 +1,69 @@
+- Start Date: 2019-07-01
+- Relevant Team(s): Ember.js
+- RFC PR: 
+- Tracking: 
+
+# Route Data Loading for Components
+
+## Summary
+
+[#499 An alternative to Controllers](https://github.com/emberjs/rfcs/pull/499) proposes an API for rendering a component from a route. This RFC builds on it by proposing an API to pass multiple arguments to a component from a route. It's based off the [design proposed by @chadhietala](https://gist.github.com/chadhietala/50b977a7d3476069892d351c65af418c) and much of the text is lifted from there.
+
+## Motivation
+
+The `model()` hook on routes is confusing, because it implies that only one piece of data will be fetched, and it uses the term "model" that is more typically associated with backend development. We should have a better name, and an API that allows data to be passed from a route into a component as multiple named arguments.
+
+## Detailed design
+
+This RFC proposes adding a `load()` hook to `Route`. The route is in the `loading` state while this hook is run. The keys of the object returned from `load()` become the named arguments to the component.
+
+```js
+// routes/profile.js
+export default class extends Route {
+  @service store;
+  async load({ profile_id }) {
+    return {
+      profile: await this.store.findRecord('profile', profile_id),
+    }
```

Comment:
Can you update this to use [`RSVP.hash`](http://api.emberjs.com/ember/3.10/functions/rsvp/hash) and multiple models?

```js
return await hash({
  profile: this.store.findRecord('profile', profile_id),
  comments: this.store.query('comments', { profile_id })
});
```

Reply:
What do you see as the advantage of using `RSVP.hash` over native JavaScript?

Reply:
Jan's example above will make the network requests run in parallel, while...

```js
return {
  profile: await this.store.findRecord('profile', profile_id),
  comments: await this.store.query('comments', { profile_id })
}
```

... will make them run sequentially.

Reply:
Hang on now, that's not totally fair. The native js way to do parallel is:

```ts
let [profile, comments] = await Promise.all([
  this.store.findRecord('profile', profile_id),
  this.store.query('comments', { profile_id })
)];

return { profile, comments };
```

Reply:
Sure. I was not saying it's not possible to do natively, just not in the way the original example was. :)

Reply:
I think we should try to stick to native JS if possible, but the `Promise.all` approach feels verbose. What it we just let the user provide an object, and if any of the values are promises, then the `load()` hook will wait for them to resolve before rendering the component (much like the `model()` hook currently does)?

Reply:
@michaelrkn, does the model hook do that if you declare it as async?

```ts
async model(params) {
  let requiredData = this.store.query('something',  { params });

  return {
     foo: requiredData.promiseReturningThing,
     bar: requiredData.someOtherPromise
  }
}
```

I think RSVP.hash has a special behavior, because it wraps the object in a _single_ promise.
I don't think model works with an object of promises.

Really, what we need is native Promise to implement `.hash`

Reply:
I'm thinking that we just wrap whatever is returned from `load()` in an RSVP hash (or a native equivalent), something like this:

```js
function load() {
  return {
    foo: this.store.findAll('foo'),
    bar: "bar data"
  };
}

async function whateverCallsLoad() {
  let data = load();
  await hash(data);
  // render component, etc.
}
```

Then the user doesn't have to get hung up on how to deal with promises - we just take care of them in the implementation (much like how the `model` hook currently does).

Reply:
I guess you can try it out here: https://github.com/emberjs/ember.js/pull/18173

relevant change here: https://github.com/emberjs/ember.js/pull/18173/files#diff-7e252e0880a2b1de772cbcabaf3ddfb4R1195

Reply:
I started to update this PR, but I realized there's a use case this doesn't help with. For simply loading data, we're fine:

```js
load({ profile_id }) {
  return {
    profile: this.store.findRecord('profile', profile_id)
  }
}
```

And for redirecting before loading data we're fine:

```js
@service router;
@service user
load({ profile_id }) {
  if (this.user.loggedIn) {
    return {
      profile: this.store.findRecord('profile', profile_id)
    };
  } else {
    this.router.transitionTo('login');
  }
}
```

But for redirecting after loading data, you have to break the pattern:

```js
load({ profile_id }) {
  let data = {
    profile: this.store.findRecord('profile', profile_id)
  };
  if (data.profile.admin) { // this doesn't work, as the promise isn't resolved
    return data;
  } else {
    this.router.transitionTo('accessDenied');
  }
}
```

I'm a bit torn. We could consider this an edge case, and have the Guides tell people to use RSVP.hash in this situation (or native Promise.all). Or we could tell people to use RSVP.hash in all situations, which would be more explicit and make it straightforward for people to figure out how to handle this case, but which would require everybody learning about an Ember-specific tool. Thoughts?

Reply:
sounds like your a step away from bringing back before/after model :p

Reply:
Ha, it crossed my mind. I don't think beforeModel solves anything. I think I'd prefer RSVP.hash or Promise.all to afterModel, but I'm open to other opinions.

Reply:
Okay, after thinking on this for a bit, I'm right back where we started - I think RSVP.hash is the way to go. It doesn't have any "magic" that the user needs to learn, and it's got a concise syntax that will make sense to most people. I'd like to use a more descriptive name than "hash", though, that gets away from it's Ruby roots and better describes what it's doing. I'm going to go with `resolveKeys` for now, but I'm open to other ideas.

Reply:
Ruby probably got it from https://en.wikipedia.org/wiki/Hash_table 😉

Reply:
FWIW the broader (non-RSVP) JS community uses [`p-props`](https://github.com/sindresorhus/p-props).